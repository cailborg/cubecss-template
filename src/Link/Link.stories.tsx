import { Story, Meta } from "@storybook/react/types-6-0";
import { Link, LinkProps } from "./Link";
import { Text } from "../Text/Text";

export default {
  title: "Components/Link",
  component: Link,
  argTypes: {
    children: { control: "text" },
  },
} as Meta;

export const Standard: Story<LinkProps> = (args) => <Link {...args}></Link>;
Standard.args = { children: 'Example Text', size: "medium", target: "_self", inline: true };


export const Inline: Story<LinkProps> = (args) => <Text variant="body-1" className="text-sunrise-700"><Link {...args}></Link></Text>;
Inline.args = { children: 'Example Text', size: "medium", target: "_self", inline: true};