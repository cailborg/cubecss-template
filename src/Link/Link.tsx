import React from "react";
import { Subtract } from "utility-types";
import styled from "@emotion/styled";
import { FocusRing } from "@react-aria/focus";

export interface LinkProps {
    href?: string;
    target?: "_blank" | "_self" | "_parent" | "_top" | "framename";
    children: string;
    trackingId?: string;
    onClick?: (event: React.MouseEvent<HTMLElement, MouseEvent>) => void;
    inline?: boolean;
    size: "legal" | "small" | "medium";
}

const getSize = (size: string) => {
    switch (size) {
        case "legal":
            return `var(--lp-font-size-link-legal)`;
        case "small":
            return `var(--lp-font-size-link-sm)`;
        case "medium":
            return `var(--lp-font-size-link-md)`;
        default:
            return `var(--lp-font-size-link-sm)`;
    }
};

const getLineHeight = (size: string) => {
    switch (size) {
        case "legal":
            return `var(--lp-line-height-link-legal)`;
        case "small":
            return `var(--lp-line-height-link-sm)`;
        case "medium":
            return `var(--lp-line-height-link-md)`;
        default:
            return `var(--lp-line-height-link-sm)`;
    }
};

const getLetterSpacing = (size: string) => {
    switch (size) {
        case "legal":
            return `var(--lp-letter-spacing-link-legal)`;
        case "small":
            return `var(--lp-letter-spacing-link-sm)`;
        case "medium":
            return `var(--lp-letter-spacing-link-md)`;
        default:
            return `var(--lp-letter-spacing-link-sm)`;
    }
};

const StyledLink = styled.a<{ inline: boolean; size: string }>`
    color: ${(props) =>
        props.inline ? `currentColor` : `rgba(var(--lp-colors-primary), 1)`};
    font-family: ${(props) =>
        props.inline ? `inherit` : `var(--lp-fonts-body-1)`};
    font-weight: inherit;
    cursor: pointer;
    text-decoration-line: underline;
    font-size: ${(props) => (props.inline ? `inherit` : getSize(props.size))};
    line-height: ${(props) =>
        props.inline ? `inherit` : getLineHeight(props.size)};
    letter-spacing: ${(props) =>
        props.inline ? `inherit` : getLetterSpacing(props.size)};
    &:hover {
        text-decoration-line: none;
        color: ${(props) =>
            props.inline
                ? `currentColor`
                : `rgba(var(--lp-colors-primary), 0.8)`};
    }
    &:focus {
        outline: none;
    }
    &.focus-ring {
        border-radius: var(--lp-radii-sm);
        box-shadow: var(--lp-focus-ring);
    }
`;

export class Link extends React.Component<LinkProps> {
    static withRoutingComponent = <P extends {}>(
        RoutingComponent: React.ComponentType<P>
    ) => {
        const Link: React.FC<Subtract<P, {}>> = (props) => (
            <RoutingComponent {...(props as P)} />
        );

        return Link;
    };

    render() {
        
        const {
            href = "#",
            target,
            trackingId,
            children,
            onClick,
            inline = false,
            size,
            ...other
        } = this.props;

        const onClickHandler = (event: React.MouseEvent<HTMLAnchorElement>) => {
            event.preventDefault();
            onClick && onClick(event);
        };

        return (
            <FocusRing focusRingClass="focus-ring">
                <StyledLink
                    inline={inline}
                    data-tracking-id={trackingId}
                    href={href}
                    target={target}
                    onClick={onClick ? onClickHandler : undefined}
                    rel="noopener noreferrer"
                    size={size}
                    {...other}
                >
                    {children}
                </StyledLink>
            </FocusRing>
        );
    }
}
