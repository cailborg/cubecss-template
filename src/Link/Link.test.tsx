import React from "react";
import { Link } from "./Link";
import renderer from "react-test-renderer";

it("renders correctly", () => {
  const tree = renderer.create(<Link size="medium">Hello</Link>).toJSON();
  expect(tree).toMatchSnapshot();
});
