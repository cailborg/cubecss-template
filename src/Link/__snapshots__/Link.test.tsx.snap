// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<a
  className=" css-k86ds9"
  href="#"
  onBlur={[Function]}
  onFocus={[Function]}
  rel="noopener noreferrer"
  size="medium"
>
  Hello
</a>
`;
